### ISSUE TITLE
[console] Purge old contact entries where instances don't exist anymore or user account is deleted

### ISSUE THREAD

**Issue by Quix0r** [friendica/friendica#10021]
With referring to [this conversation](https://f.haeder.net/display/adbba6c6-9160-4793-562f-4b8408353352), it might be nice to have a maintenance console command to cleanup old (closed) accounts where there will be no response coming from. Examples for this are many quitter and GNUSocial instances.

These are only nominal members and will no longer become active (as the instances was closed by the admin). I would suggest something like `./bin/console cleanup:orphaned-contacts [--yes] [--verbose]` which then can be setup as a cron job.

Another cleanup job could be users (`(g)contact` table?) from blocked servers (with their posts/comments) but that's another story maybe.

**Proposed function**:
 - without parameters, the script remains silent and checks if a in valid response is returned for each `contact.notify` entry.
 - `--verbose` causes each checked entry being shown plus the result (OK, FAILED)
 - `--yes` answers the then following question "Do you want to delete X orphaned contacts from your instance?" with `Y`

**Comment by bkil** [comment]
It can be seen that the `contact` and `apcontact` tables can grow without bounds. Perhaps we could hook something like this up into the existing `dbclean_*` mechanism and with similar config options?

**Comment by Quix0r** [comment]
Yes, that would become a problem one day, when your database grows and grows ...

**Comment by Xipherisroot** [comment]
I found this because I have a single user instance with /very/ light use, I only follow 42 people who are also not very active, my apcontacts table is at almost 100,000 rows and there's no reasonable way to manually remove the entries it seems.

I have my node set to "7 days" for remote items, unclaimed items, raw conversation data, etc... I also have an item-uri table that's 172,000 rows, 56,000 photo rows, 62,000 post-tag rows, etc...

Having a 530MB database, 1.5gb in the storage folder (I have almost /NO/ uploads of my own! 13 tiny photos... like maybe 6mb?) with nearly 1,000,000 rows total in the database for making maybe a post a day on average and only following 42 people of which maybe 6-7 are active and with the same volume is just unreasonable and unsustainable.

Is there really no mechanism to remove... like anything? I want to remove any 'contact' that I'm not following, I don't want to pull anything from any servers that I'm not following someone on, I don't want to keep files and photos for remote content beyond a certain age locally, etc.

**Comment by johnassel** [comment]
Is there a solution for this? Because I have several thousand contacts which I never interacted with (single user instance, too) - and these contacts are filling the database slowly. The contacts table has nearly 1.5 million entries now.

**Comment by annando** [comment]
Currently there is no progress here. But although 1.5 million sounds huge, this shouldn't harm your system.

**Comment by Xipherisroot** [comment]
> Is there a solution for this? Because I have several thousand contacts which I never interacted with (single user instance, too) - and these contacts are filling the database slowly. The contacts table has nearly 1.5 million entries now.

I had to stop using Friendica sadly since the database ended up getting massive (dozens of gigs for a single user instance...) and there's no way of cleaning it up, went to Mastodon where there's clean up tools built right into the management tools, no problems since... I really liked Friendica but its design ethos of "everything can grow forever always, nothing ever gets removed/cleaned" made it impossible to sustain in the long run :/

**Comment by johnassel** [comment]
> Currently there is no progress here. But although 1.5 million sounds huge, this shouldn't harm your system.
Yes, performance is not affected currently. Although the size of the table is over 3GB which of course affects things like backups or such - especially when several backups are kept.

My biggest tables are:
Table | Size (MB)
apcontact| 3513
contact     | 3283
gserver     | 1405

Same situation as @Xipherisroot - single user instance.
I had a look in these tables and there are contacts and servers in there which I never interacted with. Since currently no cleanup is supported is there a way to prevent these tables from filling up? Or is it possible to safely truncate the tables in order to let them rebuild?

**Comment by annando** [comment]
Depending on the admin settings, the system will try to discover contacts, see here:
![grafik](https://github.com/friendica/friendica/assets/844208/d8def70b-20c1-4ca1-b6e2-86176a829245)
What are your settings here?

**Comment by johnassel** [comment]
Ah ok - this might it be. I've disabled contact discovery - perhaps this helps.
What happens when I truncate both tables - contact and apcontact in order to get rid of the orphans? Do they get rebuilt?

**Comment by annando** [comment]
You destroy your system. Never ever delete anything in your database by hand. Since the system relies heavily on dependencies between the tables, in the best case you would receive a lot of errors because of foreign key violations. But depending on what you delete, this could also cause a deletion cascade that in the end could wipe out all your data on the system.

**Comment by johnassel** [comment]
Yes - I expected that so I didn't do that :smile: 
So there is no workaround to clean up these tables currently? Ok, then I hope they won't grow significantly anymore.

**Comment by annando** [comment]
Since you disabled this option, the growth should be much lower. Also I prepare a PR that could help as well.

**Comment by annando** [comment]
With PR #13674 admins now have got better control about server and contact updates:
![grafik](https://github.com/friendica/friendica/assets/844208/1c74a51a-8aeb-4814-8fd1-1ed2e6a26a52)
When the option "Only update contacts/servers with local data" is enabled, the system will only update contacts that local users subscribed to or that have got some local data (posts, comments). And it then only updates servers with contacts with local data. This will reduce the load on the system.

Also the explanation for the option "Discover contacts from other servers" is updated to make clear that it impacts database size and performance.

**Comment by JLuc** [comment]
Same issue here on a 6 month old friendica server : only one account, just a couple of posts alltogether and following less than 20 accounts (received a total of about 120 posts)... and the database uses 36,5 Go !

Mainly :
- storage : 32Go, 800k lines
- apcontact : 1.3Go, 500k lines
- contact : 1.1Go, 380k lines

I'll have to close the server.  It's a chance i didnt open it to friends.
